docs/book/src/developer/providers/contracts/infra-machine.md

```diff
@@ -1,118 +1,495 @@
+}
 
-	Spec InfraMachineTemplateSpec `json:"spec,omitempty"`
+type FooMachineSpec struct {
+    // See other rules for more details about mandatory/optional fields in InfraMachine spec.
+    // Other fields SHOULD be added based on the needs of your provider.
 }
 
-type InfraMachineTemplateResource struct {
-	// Standard object's metadata.
-	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
-	// +optional
-	ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
-	Spec InfraMachineSpec `json:"spec"`
+type FooMachineStatus struct {
+    // See other rules for more details about mandatory/optional fields in InfraMachine status.
+    // Other fields SHOULD be added based on the needs of your provider.
 }
 ```
 
-The CRD name of the template must also have the format produced by `sigs.k8s.io/cluster-api/util/contract.CalculateCRDName(Group, Kind)`.
+For each InfraMachine resource, you MUST also add the corresponding list resource.
+The list resource MUST be named as `<InfraMachine>List`.
+
+```go
+// +kubebuilder:object:root=true
+
+// FooMachineList contains a list of foomachines.
+type FooMachineList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []FooMachine `json:"items"`
+}
+```
 
-### List Resources
+### InfraMachine: provider ID
 
-For any resource, also add list resources, e.g.
+Each Machine needs a provider ID to identify the Kubernetes Node that runs on the machine. 
+Node's Provider id  MUST surface on `spec.providerID` in the InfraMachine resource.
 
 ```go
-//+kubebuilder:object:root=true
+type FooMachineSpec struct {
+    // ProviderID must match the provider ID as seen on the node object corresponding to this machine.
+	// For Kubernetes Nodes running on the Foo provider, this value is set by the corresponding CPI component 
+	// and it has the format docker:////<vm-name>. 
+    // +optional
+    ProviderID *string `json:"providerID,omitempty"`
+    
+    // See other rules for more details about mandatory/optional fields in InfraMachine spec.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+```
+
+Once `spec.providerID` is set on the InfraMachine resource and the [InfraMachine initialization completed],
+the Cluster controller will bubble up this info in Machine's `spec.providerID`.
+
+### InfraMachine: failure domain
+
+In case you are developing an infrastructure provider which has a notion of failure domains where machines should be
+placed in, the InfraMachine resource MUST comply to the value that exists in the `spec.failureDomain` field of the Machine
+(in other words, the InfraMachine MUST be placed in the failure domain specified at Machine level).
+
+Please note, that for allowing a transparent transition from when there was no failure domain support in Cluster API
+and InfraMachine was authoritative WRT to failure domain placement (before CAPI v1.3.0),
```

Comment:
Sorry missed this the first time. Is this actually v1.3 or should it be v0.3? (same on the issue)